### ISSUE TITLE
Problem with the new version of lark-parser (0.7.0)

### ISSUE THREAD

**Issue by M40V** [populse/populse_db#33]
An error was added when using terminal weights with dynamic earley in the newest version of lark-parser. The advanced search of populse_mia doesn't work with this version (0.7.0).

[Link](https://github.com/lark-parser/lark/commit/f67eead0b4b22898fed8efd646efda975b9b2cf5) to the commit of the version causing issue and [link](https://github.com/lark-parser/lark/issues/324) to the discussion about this commit.

**Comment by servoz** [comment]
I just assigned to you, @sapetnioc ,  this issue, because I think you wrote the populse_db part about lark. So I think you'll be the most effective on this subject and of course it's only a proposal!

**Comment by servoz** [comment]
I reopen this issue since the CI show errors with the current grammar and the LARL parser !

**Comment by servoz** [comment]
So to resume, with:

1.  _This the former state_ - lark_parser Version 0.6.7; [grammar with weights on terminals](https://github.com/populse/populse_db/blob/master/python/populse_db/filter.py#L27); default parser (Earley):
    - mia crash for some search (for example a string starting with double quotes gives UnexpectedCharacters exception: No terminal defined for '%').
    - test.py gives no errors.
    - Weakness: the search in the data browser could be very long (depending of the number of visualized tags - 10s with 14 visualized tags and 2 documents* -)

2.  lark_parser Version 0.7.1; [grammar with weights on terminals](https://github.com/populse/populse_db/blob/master/python/populse_db/filter.py#L27); default parser (Earley):
    - mia crash for any search, with ValueError exception (Dynamic Earley doesn't support weights on terminals)
    - test.py gives errors
    - no time test done !

3.  lark_parser Version 0.7.1; [grammar without weights on terminals](https://github.com/populse/populse_db/blob/master/python/populse_db/filter.py#107); default parser (Earley):
    - mia crash for some search (for example a string starting with double quotes gives UnexpectedCharacters exception: No terminal defined for '%')
    - test.py gives errors
    - Weakness: the search in the data browser could be very very long (depending of the number of visualized tags - 300s with 14 visualized tags and 2 documents* -)

4. lark_parser Version 0.6.7; [grammar with weights on terminals](https://github.com/populse/populse_db/blob/master/python/populse_db/filter.py#L27) or [grammar without weights on terminals](https://github.com/populse/populse_db/blob/master/python/populse_db/filter.py#107); parser == larl:
    - mia crash for any search, with AssertionError exception (LALR doesn't yet support prioritization)
    - test.py gives failures (AssertionError: LALR doesn't yet support prioritization)
    - no time test done !

5. lark_parser Version 0.7.1;  [grammar with weights on terminals](https://github.com/populse/populse_db/blob/master/python/populse_db/filter.py#L27) or [grammar without weights on terminals](https://github.com/populse/populse_db/blob/master/python/populse_db/filter.py#107);  parser == larl:
    - mia crash for some search (for example a string starting with double quotes gives UnexpectedCharacters exception: No terminal defined for '%')
    - test.py gives errors.
    - strong point: very fast.

Thus, it seems that the former state (used before the opening of this issue, case 1) already produced mia crashes (with queries that we had not tested in mia and that did not produce errors in unit tests!) The switch to version 0.7.1 of lark_parser no longer allows weights on terminals. But, beyond this problem, the Earley parser potentially generates a very high execution time. This execution time problem does not seems to exist with the LARL parser.

Except for the unit tests that give errors in case 5, case 1 and case 5 seem actually very similar in mia ... but case 5 seems mush faster.

It would therefore be interesting to switch to the version > 0.7 of lak_parser and to use the LARL parser. I think we should review the grammar in order to never produce crashes in mia or errors during unit tests.

Not having more time to devote to this issue at the moment, I put the project back as before (case 1) until I find some time to definitively (hoping that!) fix this issue.
-------
*Done with the same data set and conditions.

**Comment by sapetnioc** [comment]
I hope I resolved this issue. All tests are passing with lark 0.7.1 and `lalr` parser. The main problem was a probable difference in the usage of  `.` (dot) between `earley` parser (our former choice, Lark default parser) and  `lalr` parser. In the first one, `.` match any character but in `lalr`, it does not match `\n` (line break). I had to redefine `QUOTED_STRING` (instead of importing it from Lark) in order to explicitly support strings containing line breaks (there are such strings in our tests). There was other issues after this change that I fixed by setting some weights on terminals.
